Source/Core/Core/IOS/USB/Emulated/Infinity.cpp

```diff
@@ -0,0 +1,978 @@
+      0x10da7a5a, 0x67dd4acc, 0xf9b9df6f, 0x8ebeeff9, 0x17b7be43, 0x60b08ed5, 0xd6d6a3e8,
+      0xa1d1937e, 0x38d8c2c4, 0x4fdff252, 0xd1bb67f1, 0xa6bc5767, 0x3fb506dd, 0x48b2364b,
+      0xd80d2bda, 0xaf0a1b4c, 0x36034af6, 0x41047a60, 0xdf60efc3, 0xa867df55, 0x316e8eef,
+      0x4669be79, 0xcb61b38c, 0xbc66831a, 0x256fd2a0, 0x5268e236, 0xcc0c7795, 0xbb0b4703,
+      0x220216b9, 0x5505262f, 0xc5ba3bbe, 0xb2bd0b28, 0x2bb45a92, 0x5cb36a04, 0xc2d7ffa7,
+      0xb5d0cf31, 0x2cd99e8b, 0x5bdeae1d, 0x9b64c2b0, 0xec63f226, 0x756aa39c, 0x026d930a,
+      0x9c0906a9, 0xeb0e363f, 0x72076785, 0x05005713, 0x95bf4a82, 0xe2b87a14, 0x7bb12bae,
+      0x0cb61b38, 0x92d28e9b, 0xe5d5be0d, 0x7cdcefb7, 0x0bdbdf21, 0x86d3d2d4, 0xf1d4e242,
+      0x68ddb3f8, 0x1fda836e, 0x81be16cd, 0xf6b9265b, 0x6fb077e1, 0x18b74777, 0x88085ae6,
+      0xff0f6a70, 0x66063bca, 0x11010b5c, 0x8f659eff, 0xf862ae69, 0x616bffd3, 0x166ccf45,
+      0xa00ae278, 0xd70dd2ee, 0x4e048354, 0x3903b3c2, 0xa7672661, 0xd06016f7, 0x4969474d,
+      0x3e6e77db, 0xaed16a4a, 0xd9d65adc, 0x40df0b66, 0x37d83bf0, 0xa9bcae53, 0xdebb9ec5,
+      0x47b2cf7f, 0x30b5ffe9, 0xbdbdf21c, 0xcabac28a, 0x53b39330, 0x24b4a3a6, 0xbad03605,
+      0xcdd70693, 0x54de5729, 0x23d967bf, 0xb3667a2e, 0xc4614ab8, 0x5d681b02, 0x2a6f2b94,
+      0xb40bbe37, 0xc30c8ea1, 0x5a05df1b, 0x2d02ef8d};
+
+  // Calculate it
+  u32 ret = 0x00000000;
+  for (u32 i = 0; i < size; ++i)
+  {
+    u8 index = u8(((ret & 0xFF) ^ buffer[i]));
+    ret = u32(((ret >> 8) ^ CRC32_TABLE[index]));
+  }
+
+  return ret;
+}
+
+std::string InfinityBase::LoadFigure(u8* buf, File::IOFile in_file, u8 position)
+{
+  std::lock_guard lock(m_infinity_mutex);
+  u8 order_added;
+
+  std::vector<u8> sha1_calc = {SHA1_CONSTANT.begin(), SHA1_CONSTANT.end() - 1};
+  for (int i = 0; i < 7; i++)
+  {
+    sha1_calc.push_back(buf[i]);
+  }
+
+  std::array<u8, 16> infinity_aes_key = GenerateInfinityFigureKey(sha1_calc);
+
+  std::unique_ptr<Common::AES::Context> ctx =
+      Common::AES::CreateContextDecrypt(infinity_aes_key.data());
+  std::array<u8, 16> infinity_decrypted_block = {};
+  ctx->CryptIvZero(&buf[16], infinity_decrypted_block.data(), 16);
+
+  u16 number = u16(infinity_decrypted_block[2]) << 8 | u16(infinity_decrypted_block[3]);
+  DEBUG_LOG_FMT(IOS_USB, "Toy Number: {}", number);
+
+  InfinityFigure& figure = m_figures[position];
+
+  figure.inf_file = std::move(in_file);
+  memcpy(figure.data.data(), buf, figure.data.size());
+  figure.present = true;
+  if (figure.order_added == 255)
+  {
+    figure.order_added = m_figure_order;
+    m_figure_order++;
+  }
+  order_added = figure.order_added;
+
+  // In the added/removed response, position needs to be 1 for the hexagon, 2 for Player 1 and
+  // Player 1's abilities, and 3 for Player 2 and Player 2's abilities. Abilities are in positions
+  // > 2 in the UI (3/5 for player 1, 4/6 for player 2) so decrement the position until < 2.
+
+  while (position > 2)
+    position -= 2;
+
+  position = position + 1;
```

Comment:
This code appears twice with a big comment next to both entries; probably worth refactoring that into its own helper function.